File: haushaltsbuch-api/src/main/java/com/github/haschi/haushaltsbuch/api/_Aggregatkennung.java
package com.github.haschi.haushaltsbuch.api;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import org.immutables.value.Value;

import java.util.UUID;

@Value.Immutable
@Eingehüllt
@JsonSerialize(as = Aggregatkennung.class)
@JsonDeserialize(as = Aggregatkennung.class)
public abstract class _Aggregatkennung extends Umhüller<UUID>
{
    public static Aggregatkennung neu() {
        return Aggregatkennung.of(UUID.randomUUID());
    }

    public static Aggregatkennung of(final String id)
    {
        return Aggregatkennung.of(UUID.fromString(id));
    }
}

File: haushaltsbuch/src/main/kotlin/org/github/haschi/haushaltsbuch/modell/core/values/Aggregatkennung.kt
package org.github.haschi.haushaltsbuch.modell.core.values

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.SerializerProvider
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import com.fasterxml.jackson.databind.deser.std.StdScalarDeserializer
import com.fasterxml.jackson.databind.ser.std.StdSerializer
import java.io.IOException

import java.util.UUID

@JsonSerialize(using = Aggregatkennung.AggregatkennungSerialisierer::class)
@JsonDeserialize(using = Aggregatkennung.AggregatkennungDeserialisierer::class)
data class Aggregatkennung(val id: UUID) {

    class AggregatkennungDeserialisierer : StdScalarDeserializer<Aggregatkennung>
    {
        constructor() : super(Aggregatkennung::class.java)

        constructor(vc: Class<*>) : super(vc)

        @Throws(IOException::class)
        override fun deserialize(
                jsonParser: JsonParser,
                deserializationContext: DeserializationContext): Aggregatkennung
        {

            val uuid = jsonParser.text

            return Aggregatkennung(UUID.fromString(uuid))
        }
    }

    class AggregatkennungSerialisierer @JvmOverloads constructor(t: Class<Aggregatkennung>? = null) : StdSerializer<Aggregatkennung>(t) {

        @Throws(IOException::class)
        override fun serialize(
                value: Aggregatkennung,
                jgen: JsonGenerator,
                provider: SerializerProvider) {
            jgen.writeString(value.id.toString())
        }
    }

    companion object {
        fun neu(): Aggregatkennung
        {
            return Aggregatkennung(UUID.randomUUID())
        }

        fun aus(id: String): Aggregatkennung
        {
            return Aggregatkennung(UUID.fromString(id))
        }
    }
}
